File: app/server/utils/filter.ts
/**
 * 敏感信息过滤工具
 * 提取自 routes/logs.ts 和 routes/docker.ts 的重复逻辑
 */

import config from './config';

let FILTER_SENSITIVE = process.env.FILTER_SENSITIVE !== 'false';

/**
 * 过滤内容中的敏感信息
 * 将匹配敏感模式的内容替换为 [FILTERED]
 */
export function filterSensitiveInfo(content: string): string {
    if (!FILTER_SENSITIVE) return content;
    if (!content || typeof content !== 'string') return content;
    let filtered = content;
    for (const pattern of config.sensitivePatterns) {
        filtered = filtered.replace(pattern, '[FILTERED]');
    }
    return filtered;
}

/**
 * 获取敏感信息过滤状态
 */
export function isFilterEnabled(): boolean {
    return FILTER_SENSITIVE;
}

/**
 * 设置敏感信息过滤状态
 */
export function setFilterEnabled(enabled: boolean): void {
    process.env.FILTER_SENSITIVE = enabled ? 'true' : 'false';
    FILTER_SENSITIVE = enabled;
}

File: app/server/internal/utils/filter.go
package utils

import (
	"regexp"
	"sync"

	"github.com/sushazhi/fnos-logmanager/internal/config"
)

var (
	filterEnabled = true
	filterMu      sync.RWMutex
	compiled      []*regexp.Regexp
	compileOnce   sync.Once
)

func compilePatterns() {
	compileOnce.Do(func() {
		patterns := config.Get().SensitivePatterns
		compiled = make([]*regexp.Regexp, 0, len(patterns))
		for _, p := range patterns {
			if re, err := regexp.Compile(p); err == nil {
				compiled = append(compiled, re)
			}
		}
	})
}

// IsFilterEnabled returns whether the sensitive info filter is enabled.
func IsFilterEnabled() bool {
	filterMu.RLock()
	defer filterMu.RUnlock()
	return filterEnabled
}

// SetFilterEnabled enables or disables the sensitive info filter.
func SetFilterEnabled(enabled bool) {
	filterMu.Lock()
	defer filterMu.Unlock()
	filterEnabled = enabled
}

// FilterSensitiveInfo filters sensitive information from a string.
func FilterSensitiveInfo(content string) string {
	compilePatterns()

	filterMu.RLock()
	enabled := filterEnabled
	filterMu.RUnlock()

	if !enabled {
		return content
	}

	for _, re := range compiled {
		content = re.ReplaceAllString(content, "[FILTERED]")
	}
	return content
}
